### PR TITLE
Feature/h2 db support

### PR DESCRIPTION
**Pull request by rodionovsasha**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ Of course, all the test data could be defined as a SQL script(or scripts), so wh
 
 Because describing test data in SQL constructions is inconvenient:
  
-* It it hard to match values to columns:
+* It is hard to match values to columns:
```
Comment (vkorobkov) [inline README.md:21]: :+1:

File: README.md [hunk 2]
```diff
@@ -43,7 +43,7 @@ last. That means poor developer has to remember the whole tables hierarchy.
 ## JFixtures way
 
 * Human readable test data description with a set of yaml files
-* SQL scrip as result
+* SQL script as a result
```
Comment (vkorobkov) [inline README.md:46]: :+1:

File: src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.nio.file.Path
+
+import static com.github.vkorobkov.jfixtures.sql.DataSourceUtil.sql
+
+class H2IntegrationTest extends Specification implements YamlVirtualFolder {
+    Path tmpFolderPath = unpackYamlToTempFolder("default.yml")
+
+    void setupSpec() {
+        sql.execute("CREATE TABLE IF NOT EXISTS users (ID INT PRIMARY KEY, NAME varchar(50) DEFAULT NULL, AGE INT DEFAULT NULL)")
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: Should we try to drop the table(if exists) before we create a new one?

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.github.vkorobkov.jfixtures.sql
+
+import org.apache.commons.dbcp2.BasicDataSource
+import spock.lang.Shared
+
+import javax.sql.DataSource
+
+class DataSourceUtil {
+    private static final String DATASOURCE_URL = "jdbc:h2:mem:jfixturesDb;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false"
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: `String` type could be removed - groovy understands

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    private static DataSource dataSource
+
+    @Shared
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: So we need `@Shared` given the field is already static?

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    public static groovy.sql.Sql sql = new groovy.sql.Sql(getDataSource())
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: Take a look at `groovy.lang.Lazy` - probably we can just do something like:
```
@Lazy
static Sql sql = {
    def dataSource = new BasicDataSource()
    dataSource.url = DATASOURCE_URL
    new Sql(dataSource)
}()
```

or even:
```
@Lazy
static def sql = new Sql(new BasicDataSource(url: DATASOURCE_URL))
```

File: src/test/resources/integration/h2_integration_test/default.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+"users.yml":
```
Comment (vkorobkov) [inline src/test/resources/integration/h2_integration_test/default.yml:1]: Let's add a few more rows?

File: src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    void cleanup() {
+        tmpFolderPath.toFile().deleteDir()
+    }
+
+    def "h2 insert test"() {
+        when:
+        sql.execute(JFixtures.h2(tmpFolderPath as String).asString())
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: I'd create a new groovy `trait` like that and extend H2 tests from it:
```
trait H2Test extends YamlVirtualFolder {
   Sql getSql() { DataSourceUtil.sql }
   def executeFixtures(folderPath) {
      sql.execute(JFixtures.h2(folderPath as String).asString())
   } 
}
```

**Review by vkorobkov — CHANGES_REQUESTED** [review]
Looks good though some polishing around is required

File: src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy (hunk 3 above)
Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: done

File: src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy (hunk 8 above)
Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: done

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy (hunk 4 above)
Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: done

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy (hunk 5 above)
Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: Hm, not really and can be removed.. But @Shared is more for readers to show them that this variable is the same for all feature methods.. making tests more readable.

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy (hunk 6 above)
Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy]: wonderful code!

File: src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.github.vkorobkov.jfixtures.sql
+
+import org.apache.commons.dbcp2.BasicDataSource
+
+class DataSourceUtil {
+    private static final DATASOURCE_URL = "jdbc:h2:mem:jfixturesDb;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false"
+
+    @Lazy
+    static def sql = new groovy.sql.Sql(new BasicDataSource(url: DATASOURCE_URL))
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/sql/DataSourceUtil.groovy:9]: :+1:

File: src/test/groovy/com/github/vkorobkov/jfixtures/util/WithResourceTest.groovy [hunk 10]
```diff
@@ -40,7 +40,7 @@ class WithResourceTest extends Specification {
         1 * resource.close()
     }
 
-    def "boadcasts the exception when the resource is null and function fails"() {
+    def "broadcasts the exception when the resource is null and function fails"() {
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/util/WithResourceTest.groovy:43]: :+1:

File: src/test/groovy/com/github/vkorobkov/jfixtures/loader/ValueTypeTest.groovy [hunk 11]
```diff
@@ -18,7 +18,7 @@ class ValueTypeTest extends Specification {
         "SQL" | ValueType.SQL
     }
 
-    def "valueOfIgnoreCase throws exceptin when wrong value is provided"() {
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/loader/ValueTypeTest.groovy]: :+1:

File: src/test/groovy/com/github/vkorobkov/jfixtures/loader/ValueTypeTest.groovy [hunk 12]
```diff
@@ -18,7 +18,7 @@ class ValueTypeTest extends Specification {
...
+    def "valueOfIgnoreCase throws exception when wrong value is provided"() {
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/loader/ValueTypeTest.groovy:21]: :+1:

File: src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+import com.github.vkorobkov.jfixtures.testutil.H2Test
+import spock.lang.Specification
+
+import java.nio.file.Path
+
+class H2IntegrationTest extends Specification implements H2Test {
+    Path tmpFolderPath = unpackYamlToTempFolder("default.yml")
+
+    void setupSpec() {
+        sql.execute("DROP TABLE IF EXISTS users")
```
Comment (vkorobkov) [inline src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: Does it create a table before every test method of before all the methods?

Comment (rodionovsasha) [reply src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: def setupSpec() {} //run before the first feature method.

def setup() {}          // run before every feature method

Comment (vkorobkov) [reply src/test/groovy/com/github/vkorobkov/jfixtures/integration/H2IntegrationTest.groovy]: Maybe we'll need some pretty API in `H2Test` trait for dropping and creating a new table.

Would be nice just to call a method like: `recreateTable("users", [ID: "INT PRIMARY KEY", NAME: "varchar(50) DEFAULT NULL", AGE: "GE INT DEFAULT NULL"])`

**Review by vkorobkov — APPROVED** [review]

**Pull request merged**